File: mobile/src/main/java/com/windscribe/mobile/newsfeedactivity/NewsFeedPresenterImpl.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.mobile.newsfeedactivity;


import static com.windscribe.vpn.Windscribe.appContext;

import com.windscribe.mobile.adapter.NewsFeedAdapter;
import com.windscribe.vpn.ActivityInteractor;
import com.windscribe.vpn.api.response.PushNotificationAction;
import com.windscribe.vpn.errormodel.WindError;
import com.windscribe.vpn.localdatabase.tables.NewsfeedAction;
import com.windscribe.vpn.localdatabase.tables.WindNotification;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class NewsFeedPresenterImpl implements NewsFeedPresenter, NewsFeedListener {

    private final String TAG = "news_feed_p";

    private NewsFeedAdapter mAdapter;

    private final ActivityInteractor mFeedInteractor;

    private final NewsFeedView mNewsFeedView;

    private final Logger mPresenterLog = LoggerFactory.getLogger(TAG);

    @Inject
    public NewsFeedPresenterImpl(NewsFeedView mNewsFeedView, ActivityInteractor activityInteractor) {
        this.mNewsFeedView = mNewsFeedView;
        this.mFeedInteractor = activityInteractor;
    }

    @Override
    public void onDestroy() {
        mFeedInteractor.getCompositeDisposable();
        if (!mFeedInteractor.getCompositeDisposable().isDisposed()) {
            mFeedInteractor.getCompositeDisposable().dispose();
        }
    }

    @Override
    public void init(boolean showPopUp, int popUpId) {
        //Set news feed alert to false
        mNewsFeedView.showProgress("Loading");
        mFeedInteractor.getAppPreferenceInterface().setShowNewsFeedAlert(false);
        mFeedInteractor.getCompositeDisposable().add(
                mFeedInteractor.getNotifications()
                        .onErrorResumeNext(mFeedInteractor.getNotificationUpdater().update().andThen(mFeedInteractor.getNotifications()))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(notifications -> onNotificationResponse(showPopUp, popUpId, notifications),
                                this::onNotificationResponseError));
    }

    @Override
    public void onNotificationActionClick(final WindNotification windNotification) {
        NewsfeedAction newsfeedAction = windNotification.getAction();
        if (newsfeedAction != null) {
            PushNotificationAction pushNotificationAction = new PushNotificationAction(newsfeedAction.getPcpID(),
                    newsfeedAction.getPromoCode(), newsfeedAction.getType());
            if (pushNotificationAction.getType().equals("promo")) {
                mNewsFeedView.startUpgradeActivity(pushNotificationAction);
            }
        }
    }

    @Override
    public void onNotificationExpand(WindNotification windNotification) {
        mFeedInteractor.getAppPreferenceInterface()
                .saveNotificationId(String.valueOf(windNotification.getNotificationId()));
    }

    private void onNotificationResponse(boolean showPopUp, int popUpId, List<WindNotification> mNotificationList) {
        mPresenterLog.info("Loaded notification data successfully...");
        int firstItemToOpen = -1;
        for (WindNotification wn : mNotificationList) {
            boolean read = mFeedInteractor.getAppPreferenceInterface()
                    .isNotificationAlreadyShown(String.valueOf(wn.getNotificationId()));
            if (!read && firstItemToOpen == -1) {
                firstItemToOpen = wn.getNotificationId();
            }
            wn.setRead(read);
        }
        if (showPopUp) {
            mPresenterLog.debug("Showing pop up message with Id: " + popUpId);
            firstItemToOpen = popUpId;
        } else if (firstItemToOpen != -1) {
            mPresenterLog.debug("Showing unread message with Id: " + firstItemToOpen);
        } else {
            mPresenterLog.debug("No pop up or unread message to show");
        }
        mAdapter = new NewsFeedAdapter(mNotificationList, firstItemToOpen,
                NewsFeedPresenterImpl.this);
        mNewsFeedView.setNewsFeedAdapter(mAdapter);
        mNewsFeedView.hideProgress();
    }

    private void onNotificationResponseError(Throwable throwable) {
        mPresenterLog.debug("Error getting notification data. Error: " +
                WindError.getInstance().convertThrowableToString(throwable));
        mNewsFeedView.showLoadingError("Error loading news feed data...");
        mNewsFeedView.hideProgress();
    }
}

File: mobile/src/main/java/com/windscribe/mobile/newsfeedactivity/NewsFeedPresenterImpl.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */
package com.windscribe.mobile.newsfeedactivity

import com.windscribe.mobile.adapter.NewsFeedAdapter
import com.windscribe.vpn.ActivityInteractor
import com.windscribe.vpn.api.response.PushNotificationAction
import com.windscribe.vpn.errormodel.WindError.Companion.instance
import com.windscribe.vpn.localdatabase.tables.WindNotification
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import org.slf4j.LoggerFactory
import javax.inject.Inject

class NewsFeedPresenterImpl @Inject constructor(
    private val newsFeedView: NewsFeedView,
    private val interactor: ActivityInteractor
) : NewsFeedPresenter, NewsFeedListener {
    private var adapter: NewsFeedAdapter? = null
    private val logger = LoggerFactory.getLogger("news_feed_p")
    override fun onDestroy() {
        interactor.getCompositeDisposable()
        if (!interactor.getCompositeDisposable().isDisposed) {
            interactor.getCompositeDisposable().dispose()
        }
    }

    override fun init(showPopUp: Boolean, popUpId: Int) {
        //Set news feed alert to false
        newsFeedView.showProgress("Loading")
        interactor.getAppPreferenceInterface().setShowNewsFeedAlert(false)
        interactor.getCompositeDisposable().add(
            interactor.getNotifications()
                .onErrorResumeNext(
                    interactor.getNotificationUpdater().update()
                        .andThen(interactor.getNotifications())
                )
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ notifications: List<WindNotification> ->
                    onNotificationResponse(
                        showPopUp,
                        popUpId,
                        notifications
                    )
                }) { throwable: Throwable -> onNotificationResponseError(throwable) })
    }

    override fun onNotificationActionClick(windNotification: WindNotification) {
        val newsfeedAction = windNotification.action
        if (newsfeedAction != null) {
            val pushNotificationAction = PushNotificationAction(
                newsfeedAction.pcpID,
                newsfeedAction.promoCode, newsfeedAction.type
            )
            if (pushNotificationAction.type == "promo") {
                newsFeedView.startUpgradeActivity(pushNotificationAction)
            }
        }
    }

    override fun onNotificationExpand(windNotification: WindNotification) {
        interactor.getAppPreferenceInterface()
            .saveNotificationId(windNotification.notificationId.toString())
    }

    private fun onNotificationResponse(
        showPopUp: Boolean,
        popUpId: Int,
        mNotificationList: List<WindNotification>
    ) {
        logger.info("Loaded notification data successfully...")
        var firstItemToOpen = -1
        for (wn in mNotificationList) {
            val read = interactor.getAppPreferenceInterface()
                .isNotificationAlreadyShown(wn.notificationId.toString())
            if (!read && firstItemToOpen == -1) {
                firstItemToOpen = wn.notificationId
            }
            wn.isRead = read
        }
        if (showPopUp) {
            logger.debug("Showing pop up message with Id: $popUpId")
            firstItemToOpen = popUpId
        } else if (firstItemToOpen != -1) {
            logger.debug("Showing unread message with Id: $firstItemToOpen")
        } else {
            logger.debug("No pop up or unread message to show")
        }
        adapter = NewsFeedAdapter(
            mNotificationList, firstItemToOpen,
            this@NewsFeedPresenterImpl
        )
        newsFeedView.setNewsFeedAdapter(adapter!!)
        newsFeedView.hideProgress()
    }

    private fun onNotificationResponseError(throwable: Throwable) {
        logger.debug(
            "Error getting notification data. Error: " +
                    instance.convertThrowableToString(throwable)
        )
        newsFeedView.showLoadingError("Error loading news feed data...")
        newsFeedView.hideProgress()
    }
}